core/unit_test/TestTeam.hpp

```diff
@@ -1529,6 +1529,38 @@ struct TestScratchAlignment {
 
 }  // namespace
 
+namespace {
+
+template <class ExecSpace>
+struct TestTeamPolicyHandleByValue {
```

Comment:
How is that test related to the issue you are fixing?

Reply:
It's a reproducer for the issue @ndellingwood was seeing when testing with `Trilinos`.

Reply:
How is it related to whether the team handle is passed by value or not?!  :/

Reply:
https://github.com/kokkos/kokkos/pull/3957#issue-620735358

Reply:
I was able to reproduce the problem with `RelWithDebInfo` but didn't see it for `Debug` with `ICC18`. I would suspect that some optimizations lead to wrong assembly...

Reply:
It tests taking the team policy handle by value rather than by reference, in this particular test it seg faults in builds with intel without the changes in Kokkos_ScratchSpace.hpp (I saw with a Serial build, debug mode, and intel/18, did not test more broadly). 
Testing taking the team policy handle by value rather than reference may be done in other unit tests, so there may be some peculiarities about this particular case that are tickling compiler issues, for example if I replaced the lambda body of `a(i, j) += b(i, j);` with `a(i, j) += 1;` the seg fault was not present